Firebase/Core/FIRLogger.m

```diff
@@ -86,170 +64,53 @@
-      aslOptions = ASL_OPT_STDERR;
-    }
-
-#pragma clang diagnostic push
-#pragma clang diagnostic ignored "-Wdeprecated-declarations"  // asl is deprecated
-    // Initialize the ASL client handle.
-    sFIRLoggerClient = asl_open(NULL, kFIRLoggerASLClientFacilityName, aslOptions);
-
-    // Set the filter used by system/device log. Initialize in default mode.
-    asl_set_filter(sFIRLoggerClient, ASL_FILTER_MASK_UPTO(ASL_LEVEL_NOTICE));
-    sFIRLoggerDebugMode = NO;
-    sFIRAnalyticsDebugMode = NO;
-    sFIRLoggerMaximumLevel = FIRLoggerLevelNotice;
-
-    // Use the standard NSUserDefaults if it hasn't been explicitly set.
-    if (sFIRLoggerUserDefaults == nil) {
-      sFIRLoggerUserDefaults = [NSUserDefaults standardUserDefaults];
-    }
-
-    BOOL debugMode = [sFIRLoggerUserDefaults boolForKey:kFIRPersistedDebugModeKey];
-    if ([arguments containsObject:kFIRDisableDebugModeApplicationArgument]) {  // Default mode
-      [sFIRLoggerUserDefaults removeObjectForKey:kFIRPersistedDebugModeKey];
-    } else if ([arguments containsObject:kFIREnableDebugModeApplicationArgument] ||
-               debugMode) {  // Debug mode
-      [sFIRLoggerUserDefaults setBool:YES forKey:kFIRPersistedDebugModeKey];
-      asl_set_filter(sFIRLoggerClient, ASL_FILTER_MASK_UPTO(ASL_LEVEL_DEBUG));
-      sFIRLoggerDebugMode = YES;
-    }
-
-    // We should disable debug mode if we are running from App Store.
-    if (sFIRLoggerDebugMode && [FIRAppEnvironmentUtil isFromAppStore]) {
-      sFIRLoggerDebugMode = NO;
-    }
-
-    sFIRClientQueue = dispatch_queue_create("FIRLoggingClientQueue", DISPATCH_QUEUE_SERIAL);
-    dispatch_set_target_queue(sFIRClientQueue,
-                              dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0));
-
-#ifdef DEBUG
-    sMessageCodeRegex =
-        [NSRegularExpression regularExpressionWithPattern:kMessageCodePattern options:0 error:NULL];
-#endif
+    GULLoggerInitializeASL([arguments containsObject:kFIRDisableDebugModeApplicationArgument],
+                           [arguments containsObject:kFIREnableDebugModeApplicationArgument],
+                           [arguments containsObject:kFIRLoggerForceSDTERRApplicationArgument]);
   });
 }
 
 void FIRSetAnalyticsDebugMode(BOOL analyticsDebugMode) {
-  FIRLoggerInitializeASL();
-  dispatch_async(sFIRClientQueue, ^{
-    // We should not enable debug mode if we are running from App Store.
-    if (analyticsDebugMode && [FIRAppEnvironmentUtil isFromAppStore]) {
-      return;
-    }
-    sFIRAnalyticsDebugMode = analyticsDebugMode;
-    asl_set_filter(sFIRLoggerClient, ASL_FILTER_MASK_UPTO(ASL_LEVEL_DEBUG));
-  });
-}
-
-void FIRSetLoggerLevel(FIRLoggerLevel loggerLevel) {
-  if (loggerLevel < FIRLoggerLevelMin || loggerLevel > FIRLoggerLevelMax) {
-    FIRLogError(kFIRLoggerCore, @"I-COR000023", @"Invalid logger level, %ld", (long)loggerLevel);
+  if (analyticsDebugMode && [GULAppEnvironmentUtil isFromAppStore]) {
```

Comment:
This is modifying global state without a mutex or a queue. If this is intentional then you probably need a tsan annotation: `__attribute__((no_sanitize("thread")))`.